endpoints/openrtb2/video_auction.go

```diff
@@ -0,0 +1,524 @@
+	a. Load the stored impression to use as the basis for impressions generated for this pod from the configid field.
+	b. NumImps = adpoddurationsec / MIN_VALUE(allowedDurations)
+	c. Build impression array for this pod:
+		I.Create array of NumImps entries initialized to the base impression loaded from the configid.
+			1. If requireexactdurations = true, iterate over allowdDurations and for (NumImps / len(allowedDurations)) number of Imps set minduration = maxduration = allowedDurations[i]
+			2. If requireexactdurations = false, set maxduration = MAX_VALUE(allowedDurations)
+		II. Set Imp.id field to "podX_Y" where X is the pod index and Y is the impression index within this pod.
+	d. Append impressions for this pod to the overall list of impressions in the OpenRTB bid request.
+9. Call validateRequest() function from auction.go to validate the generated request.
+10. Call HoldAuction() function to run the auction for the OpenRTB bid request that was built in the previous step.
+11. Build proper response format.
+*/
+func (deps *endpointDeps) VideoAuctionEndpoint(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+
+	ao := analytics.AuctionObject{
+		Status: http.StatusOK,
+		Errors: make([]error, 0),
+	}
+
+	start := time.Now()
+	labels := pbsmetrics.Labels{
+		Source:        pbsmetrics.DemandUnknown,
+		RType:         pbsmetrics.ReqTypeORTB2Web,
+		PubID:         "",
+		Browser:       pbsmetrics.BrowserOther,
+		CookieFlag:    pbsmetrics.CookieFlagUnknown,
+		RequestStatus: pbsmetrics.RequestStatusOK,
+	}
+	numImps := 0
+	defer func() {
+		deps.metricsEngine.RecordRequest(labels)
+		deps.metricsEngine.RecordImps(labels, numImps)
+		deps.metricsEngine.RecordRequestTime(labels, time.Since(start))
+		deps.analytics.LogAuctionObject(&ao)
+	}()
+
+	isSafari := checkSafari(r)
+	if isSafari {
+		labels.Browser = pbsmetrics.BrowserSafari
+	}
+
+	lr := &io.LimitedReader{
+		R: r.Body,
+		N: deps.cfg.MaxRequestSize,
+	}
+	requestJson, err := ioutil.ReadAll(lr)
+	if err != nil {
+		errL := []error{err}
+		handleError(labels, w, errL, ao)
+		return
+	}
+
+	resolvedRequest := requestJson
+
+	//load additional data - stored simplified req
+	storedRequestId, err := getVideoStoredRequestId(requestJson)
+
+	if err != nil && deps.cfg.VideoStoredRequestRequired {
+		errL := []error{err}
+		handleError(labels, w, errL, ao)
+		return
+	}
+	if err == nil {
+		storedRequest, err := loadStoredVideoRequest(storedRequestId)
+		if err != nil {
+			errL := []error{err}
+			handleError(labels, w, errL, ao)
+			return
+		}
+
+		//merge incoming req with stored video req
+		resolvedRequest, err = jsonpatch.MergePatch(storedRequest, requestJson)
+	}
+	//unmarshal and validate combined result
+	videoBidReq, errL := deps.parseVideoRequest(resolvedRequest)
+	if len(errL) > 0 {
+		handleError(labels, w, errL, ao)
+		return
+	}
+
+	var bidReq = &openrtb.BidRequest{}
+	if deps.defaultRequest {
+		hasErr, Err := getJsonSyntaxError(deps.defReqJSON)
```

Comment:
I am not sure we need this because we are unmashalling into an empty structure. In the main code we are merging two JSON blobs and the error message was confusing as to which JSON blob was bad.

Reply:
Done